### PR TITLE
Remask tool response when it is nested in assistant span

### PR DESCRIPTION
**Pull request by shanghongsim**
# Description

PR https://github.com/oumi-ai/oumi/pull/2590 is being split into two parts. This PR contains part 1: the tool remasking logic from https://github.com/oumi-ai/oumi/pull/2590

## Background
Some chat templates render tool results inside the assistant turn (like Gemma 4), which leaves them unmasked. This results in model being trained on environment output leading to model degradation.

This seems to be a common problems faced by the community trying to use Gemma 4 for agentic cases:
[git.sethpc.xyz/Seth/gemma4-research/src/commit/91aaaa48d7275333e92f76a6b4be6c1a5cdbeb89/tooling/fine-tuning](https://git.sethpc.xyz/Seth/gemma4-research/src/commit/91aaaa48d7275333e92f76a6b4be6c1a5cdbeb89/tooling/fine-tuning)
there has not been a canonical way or fix suggested by the community for this issue

The original PR https://github.com/oumi-ai/oumi/pull/2590 suggested a two fold fix:
1. In `collators.py`: We have a mechanism that autodetects tool_response start and end if the user does not supply it and we detect that tool response is indeed nested in assistant span.
2. In `trl_data_collator_for_completion_only_lm.py`: we remask the regions between tool_response start and end, if they are supplied.

Gemma 4's chat template did indeed change between the API repo pinned version (`4742fe843c`) and the latest commit.

In the pinned version, tool response is in tool turn (refer to picture 1). In their latest version, tool response has been shifted to be with assistant (refer to picture 2).

image

image

## Design

@wizeng23 flagged that the autodetection of the tool response template has high blast radius in the original implementation. We worked together to find a few ways to derisk the process:

1. Split the probe and the remasking logic into two different PRs. The remaksing logic can merge in first to unblock critical workflows. While the probe detection logic can be tested and reviewed more thoroughly

2. Limit the activation of probe using certain checks in `collators.py`:

```
if the user has supplied the tool response prefix, apply remasking using this
for models with the known nesting behavior, we maintain a dict of the tool response prefix and just apply the remask directly
else, run the probe to autodetect the tool prefix. 
```

This PR contains the above two things.

A followup PR will be about the probe.

## Testing

- All unit and integration tests pass
- Barring an unrelated data type issue, which I fixed locally, I validated that training runs for the following configurations on exun:

Qwen3.5 + tools dataset: https://wandb.ai/shanghongsim/huggingface/runs/josebt7u
Qwen3.5 + alpaca: https://wandb.ai/shanghongsim/huggingface/runs/ode8xa6c
Gemma 4 + tools dataset: https://wandb.ai/shanghongsim/huggingface/runs/otgby2y6
Gemma 4 + alpaca: https://wandb.ai/shanghongsim/huggingface/runs/mx7b8551

The tool dataset is a realistic enterprise use case.

image

Losses and mean accuracy look reasoanble. 

## Related issues

Fixes # (issue)

## Before submitting

- [ ] This PR only changes documentation. (You can ignore the following checks in that case)
- [x] Did you read the [contributor guideline](https://github.com/oumi-ai/oumi/blob/main/CONTRIBUTING.md) Pull Request guidelines?
- [ ] Did you link the issue(s) related to this PR in the section above?
- [x] Did you add / update tests where needed?

## Reviewers

At least one review from a member of `oumi-ai/oumi-staff` is required.

> [!NOTE]
> **Liberate**
> Risk: medium

### CODE REVIEW

File: src/oumi/builders/collators.py [hunk 1]
```diff
@@ -40,6 +41,31 @@
     "Fix: provide response_template (and end_of_turn_template for "
     "all_assistant_turns) in collator_kwargs."
 )
+_SPAN_TRAIN_TARGETS = ("all_assistant_turns", "final_assistant_turn")
```
Comment (wizeng23) [inline src/oumi/builders/collators.py:51]: nit: can you add comments to all global constants explaining what they are?

File: src/oumi/core/collators/trl_data_collator_for_completion_only_lm.py [hunk 2]
```diff
@@ -144,6 +155,39 @@ def _find_pattern(seq: list[int], pattern: list[int]) -> list[int]:
...
+        end_idx: int,
+    ) -> None:
+        """Re-masks tool results that the chat template nested in an unmasked span.
+
+        Some templates (e.g. gemma-4) render tool results inside the assistant turn, so
+        the span between response_template and end_of_turn_template contains
+        environment output the model never generates.
+
+        An opener with no closer before `end_idx` masks through `end_idx`: truncation
+        can cut a block in half, and over-masking is the safe direction.
```
Comment (wizeng23) [inline src/oumi/core/collators/trl_data_collator_for_completion_only_lm.py:198]: Can you add an Args: section to the docstring? It's not clear what the inputs are

File: src/oumi/core/collators/text_completions_collator_with_padding.py [hunk 3]
```diff
@@ -48,6 +50,11 @@ def __init__(
             or ``"final_assistant_turn"``.
         end_of_turn_template: String marking the end of a turn.
             Required for ``all_assistant_turns``.
+        tool_response_template: String or token IDs opening a tool result that the
+            chat template renders inside the assistant turn (e.g. gemma-4's
+            ``<|tool_response>``). Auto-detected by ``resolve_tool_response_template``.
```
[the remark this answers is not in the document]
Comment (shanghongsim) [reply src/oumi/core/collators/text_completions_collator_with_padding.py]: removed

**Review by wizeng23 — APPROVED** [review]
Could you please document in the PR description what E2E testing you did? i.e. confirming that this works for our gemma 4 configs, and doesn't affect other models

**Review by jgreer013 — APPROVED** [review]

File: src/oumi/core/collators/trl_data_collator_for_completion_only_lm.py (hunk 2 above)
Comment (shanghongsim) [reply src/oumi/core/collators/trl_data_collator_for_completion_only_lm.py:198]: done

File: src/oumi/builders/collators.py [hunk 4]
```diff
@@ -196,6 +222,95 @@ def resolve_collator_templates(
...
+                f"'{missing}'. Tool results are only excluded from the training "
+                "loss when both markers are set.\n"
+                f"Fix: also set '{missing}' in collator_kwargs."
+            )
+        else:
+            return None  # Hand-supplied bracket wins; nothing to look up.
+
+    # 1b. Skip if on the legacy path
+    if collator_kwargs.get("train_target") not in _SPAN_TRAIN_TARGETS:
+        return None  # Only span-based masking can unmask a tool result.
```
[the remark this answers is not in the document]
Comment (shanghongsim) [reply src/oumi/builders/collators.py:377]: added test to show that providing tool response template with legacy option is inert in `test_legacy_instruction_response_accepts_the_bracket_but_never_applies_it`

File: src/oumi/core/collators/trl_data_collator_for_completion_only_lm.py [hunk 5]
```diff
@@ -263,6 +308,16 @@ def torch_call(
...
                     batch["labels"][i, :response_token_ids_end_idx] = self.ignore_index
 
+                    # The unmasked tail is one assistant turn, so it carries the same
+                    # nested-tool-result exposure as span masking.
+                    self._mask_tool_response_spans(
+                        batch,
+                        i,
+                        batch["input_ids"][i].tolist(),
+                        response_token_ids_end_idx,
+                        batch["input_ids"].shape[1],
```
[the remark this answers is not in the document]
Comment (shanghongsim) [reply src/oumi/core/collators/trl_data_collator_for_completion_only_lm.py:399]: negligible cpu savings, backprop still dominates

File: src/oumi/builders/collators.py (hunk 1 above)
Comment (shanghongsim) [reply src/oumi/builders/collators.py:51]: done

**Comment by shanghongsim** [comment]
> Could you please document in the PR description what E2E testing you did? i.e. confirming that this works for our gemma 4 configs, and doesn't affect other models

updated! FYI @wizeng23
